src/coreclr/jit/priorityqueue.h

```diff
@@ -0,0 +1,112 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+//
+// PriorityQueue: A priority queue implemented as a max-heap
+
+template <typename T, typename Compare>
+class PriorityQueue
+{
+private:
+    jitstd::vector<T> data;
+    Compare           comp;
+
+#ifdef DEBUG
+    // Returns true only if each element has a higher priority than its children.
+    bool VerifyMaxHeap() const
+    {
+        for (size_t i = 0; i < data.size(); i++)
+        {
+            const size_t leftChild  = (2 * i) + 1;
+            const size_t rightChild = leftChild + 1;
+
+            if (rightChild < data.size())
+            {
+                if (comp(data[i], data[leftChild]) || comp(data[i], data[rightChild]))
+                {
+                    return false;
+                }
+            }
+            else if (leftChild < data.size())
+            {
+                if (comp(data[i], data[leftChild]))
+                {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+#endif // DEBUG
+
+public:
+    PriorityQueue(const jitstd::allocator<T>& allocator, const Compare& compare)
+        : data(allocator)
+        , comp(compare)
+    {
+    }
+
+    const T& Top() const
+    {
+        assert(!data.empty());
+        return data.front();
+    }
+
+    bool Empty() const
+    {
+        return data.empty();
+    }
+
+    // Insert new element at the back of the vector.
+    // Then, while the new element has a higher priority than its parent, swap them.
+    void Push(const T& value)
+    {
+        size_t i = data.size();
+        data.push_back(value);
+
+        auto getParent = [](const size_t i) -> size_t {
+            return (i - 1) / 2;
+        };
+
+        for (size_t parent = getParent(i); (i != 0) && comp(data[parent], data[i]); i = parent, parent = getParent(i))
+        {
+            std::swap(data[parent], data[i]);
+        }
```

Comment:
These repeated swaps can be avoided by writing the value only once at the end, e.g. as done in the managed implementation:
https://github.com/dotnet/runtime/blob/e476b43b5cb42eb44ce23b1c7b793aa361624cf6/src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs#L713-L743

Reply:
Thanks for pointing this out. Looks like I can rework `Pop` to avoid swapping too.